libp2p/autotls/autotls.nim

```diff
@@ -0,0 +1,258 @@
+
+  return false
+
+proc new*(
+    T: typedesc[AutoTLSManager],
+    rng: ref HmacDrbgContext = newRng(),
+    acmeAccount: ref ACMEAccount = nil,
+    dnsResolver: DnsResolver = DnsResolver.new(DefaultDnsServers),
+    acmeServerURL: string = LetsEncryptURL,
+): AutoTLSManager =
+  T(
+    rng: rng,
+    managerLoop: nil,
+    cert: Opt.none(TLSCertificate),
+    certExpiry: Opt.none(Moment),
+    certReady: newAsyncEvent(),
+    acmeAccount: acmeAccount,
+    httpSession: HttpSessionRef.new(),
+    dnsResolver: dnsResolver,
+    peerInfo: Opt.none(PeerInfo),
+    bearerToken: Opt.none(string),
+    renewCheckTime: DefaultRenewCheckTime,
+    renewBufferTime: DefaultRenewBufferTime,
+    acmeServerURL: acmeServerURL,
+  )
+
+method issueCertificate(
+    self: AutoTLSManager
+): Future[void] {.base, async: (raises: [AutoTLSError, CancelledError]).} =
+  trace "Issuing new certificate"
+  var peerInfo: PeerInfo
+  try:
+    peerInfo = self.peerInfo.get
+  except:
+    raise newException(AutoTLSError, "Cannot issue new certificate: peerInfo not set")
+
+  # generate autotls domain string: "*.{peerID}.libp2p.direct"
+  let base36PeerId = encodePeerId(peerInfo.peerId)
+  let baseDomain = base36PeerId & "." & AutoTLSDNSServer
+  let domain = "*." & baseDomain
+
+  debug "Requesting ACME challenge"
+  let (dns01Challenge, finalizeURL, orderURL) =
+    await self.acmeAccount.requestChallenge(@[domain])
+
+  let keyAuthorization = base64UrlEncode(
+    @(
+      sha256.digest(
+        (
+          dns01Challenge.getJSONField("token").getStr & "." &
+          thumbprint(self.acmeAccount.key)
+        ).toByteSeq
+      ).data
+    )
+  )
+
+  var strMultiaddresses: seq[string] = @[]
+  for ma in peerInfo.addrs:
+    strMultiaddresses.add($ma)
+
+  let payload = %*{"value": keyAuthorization, "addresses": strMultiaddresses}
+  let registrationURL = AutoTLSBroker & "/v1/_acme-challenge"
+
+  debug "Sending challenge to AutoTLS broker"
+  var response: HttpClientResponseRef
+  var bearerToken: string
+  if self.bearerToken.isSome:
+    (bearerToken, response) = await peerIdAuthSend(
+      registrationURL,
+      self.httpSession,
+      peerInfo,
+      payload,
+      self.rng,
+      bearerToken = self.bearerToken,
+    )
+    discard bearerToken
+  else:
+    # authenticate, send challenge and save bearerToken for future requests
+    (bearerToken, response) = await peerIdAuthSend(
+      registrationURL, self.httpSession, peerInfo, payload, self.rng
+    )
+    self.bearerToken = Opt.some(bearerToken)
+  if response.status != 200:
+    raise newException(
+      AutoTLSError, "Failed to authenticate with AutoTLS Broker at " & AutoTLSBroker
+    )
+
+  # no need to do anything from this point forward if there are not public ip addresses on host
+  var hostPrimaryIP: IpAddress
+  try:
+    hostPrimaryIP = getPrimaryIPAddr()
+    if not isPublicIPv4(hostPrimaryIP):
+      raise newException(AutoTLSError, "Host does not have a public IPv4 address")
```

Comment:
Some projects like nwaku obtain their IP address via NAT(upmp/pmp) (look for `setupNAT` in nwaku repo), or even allow the user to hardcode an IPAddress.  It probably makes sense to have an optional attribute in AutoTLSManager so a dev can setup an IP (this should be also a parameter passed to withAutoTLSManager()) option.
If the dev using nim-libp2p sets up the optional IP, it should use that as the hostPrimaryIP, otherwise it can determine the IP just like you're doing here